WIP [KeyVault] How to fill the form of the Deploy button

sadasant opened the pull request:
(I wanted to make this PR a draft PR, but I forgot to click that button at the end)

Here's my take on a simple guide to fill the form of the deploy button. It only exists currently for KeyVault-Keys. Your feedback is appreciated! Let's make this better here so we can copy this in all of the other projects!

This is not intended to be perfect! I'd like to get as much feedback as possible, to understand what you think is valuable to add here. This is intentionally bare!

Fixes #7361
Fixes #7167

xirzec commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
 
+To fill this form:
+
+1. If you know about ARM templates, or if you're interested to know about them, we recommend you to take some time to read through the ARM template we are using for this button [here](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-sdk-for-js%2Fmaster%2Fsdk%2Fkeyvault%2Ftest-resources.json). You can read more about the concept of an  ...
```
I'm assuming the [here] link is the same thing as the deploy button above, but it feels weird to call it out with such an unhelpful name as 'here', I think 'above' and not having it be linked would suffice? Either that or actually repeat the button wholesale

sadasant replied:
Oh my bad, the here button is intended to go directly to the source of the ARM template, not to the deployment form.

sadasant replied:
Fixed this! Please read it again.

xirzec commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
 
+To fill this form:
+
+1. If you know about ARM templates, or if you're interested to know about them, we recommend you to take some time to read through the ARM template we are using for this button [here](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-sdk-for-js%2Fmaster%2Fsdk%2Fkeyvault%2Ftest-resources.json). You can read more about the concept of an  ...
```
the first line feels like it should be some text before the list instead of the first item, but I know you want people to actually read it...

sadasant replied:
I'm open to suggestions! This is meant as a conversation starter 🌞

xirzec commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+2. The subscription, resource group and location are required fields, they help specify where to create the KeyVault.
+3. On the "Base Name", you'll see `[resourceGroup().name]`. Remove the contents of this field and type any alphanumeric string. This string will be used to generate unique names for the resources that will be created.
+4. To fill the "Tenant Id" field, copy it from the one you'll be able to see once you go to the [overview page of the Azure Active Directory section of the Azure portal](https://ms.portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/Overview).
+5. To fill the "Test Application Oid", you'll first need to create an Azure Active Directory application ([here's how to](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal)). Once created, go to your application's overview page, and you'll see an "Object ID" property. This is the one that you want to copy and paste on the form.
```
instead of [here's how to] maybe just linkify the text [create an Azure Active Directory application] ?

sadasant replied:
Linkified!

HarshaNalluru commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
 
+To fill this form:
+
+1. If you know about ARM templates, or if you're interested to know about them, we recommend you to take some time to read through the [ARM template we are using for this button](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about the concept of an ARM template in our [Intro to Microsoft Azure Resource Manager Temp ...
```
Is this a reliable link? official/unofficial?

sadasant replied:
It's from Jay Gordon, published on the Azure dev.to account. I think it's official. I will ask Maggie though.

sadasant replied:
I asked around and I got this, which seems better: https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview

jeremymeng commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+
+1. If you know about ARM templates, or if you're interested to know about them, we recommend you to take some time to read through the [ARM template we are using for this button](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about the concept of an ARM template in our [Intro to Microsoft Azure Resource Manager Temp ...
+2. The subscription, resource group and location are required fields, they help specify where to create the KeyVault.
+3. On the "Base Name", you'll see `[resourceGroup().name]`. Remove the contents of this field and type any alphanumeric string. This string will be used to generate unique names for the resources that will be created.
```
styling: you might want to use **Bold** font face for portal UI elements.  I've seen the patter in many azure docs.

sadasant replied:
What would you like to make bold? Just sot that I can make sure I address your feedback properly.

jeremymeng replied:
For example, in one of the linked doc

![image](https://user-images.githubusercontent.com/7583839/74481549-cad86a00-4e67-11ea-907d-1bf8be0fd023.png)

sadasant replied:
I have used the bold markdown syntax on several words and names! Let me know if it's better now 🌞

chradek commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
 
+To fill this form:
+
+1. If you know about ARM templates, or if you're interested to know about them, we recommend you to take some time to read through the [ARM template we are using for this button](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about the concept of an ARM template in our [Azure Resource Manager templates overview](htt ...
```
Suggestion:
```suggestion
1. If you know about ARM templates, or you're interested to learn about them, we recommend you to take some time to read through the [ARM template used by the deploy button in this readme](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about ARM templates in the [Azure Resource Manager templates overview](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview).
```

chradek commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+To fill this form:
+
+1. If you know about ARM templates, or if you're interested to know about them, we recommend you to take some time to read through the [ARM template we are using for this button](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about the concept of an ARM template in our [Azure Resource Manager templates overview](htt ...
+2. The subscription, resource group and location are required fields, they help specify where to create the KeyVault.
```
```suggestion
2. The subscription, resource group and location are required fields; they help specify where to create the KeyVault.
```

chradek commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+
+1. If you know about ARM templates, or if you're interested to know about them, we recommend you to take some time to read through the [ARM template we are using for this button](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about the concept of an ARM template in our [Azure Resource Manager templates overview](htt ...
+2. The subscription, resource group and location are required fields, they help specify where to create the KeyVault.
+3. On the "Base Name", you'll see `[resourceGroup().name]`. Remove the contents of this field and type any alphanumeric string. This string will be used to generate unique names for the resources that will be created.
```
Is the Base Name a prefix? If so it might be helpful to call it out as such. Also which resources are created by this (I'm assuming just a key vault)

sadasant replied:
In this case is just one KeyVault! I'll make sure I mention that. I'll also leave this comment unresolved in the light that someone might be able to use this conversation to help them catch that their README changes need to be representative of the resources they create 👍

chradek commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+1. If you know about ARM templates, or if you're interested to know about them, we recommend you to take some time to read through the [ARM template we are using for this button](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about the concept of an ARM template in our [Azure Resource Manager templates overview](htt ...
+2. The subscription, resource group and location are required fields, they help specify where to create the KeyVault.
+3. On the "Base Name", you'll see `[resourceGroup().name]`. Remove the contents of this field and type any alphanumeric string. This string will be used to generate unique names for the resources that will be created.
+4. To fill the "Tenant Id" field, copy it from the one you'll be able to see once you go to the [overview page of the Azure Active Directory section of the Azure portal](https://ms.portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/Overview).
```
It might be easier to have the user create their App Registration in AAD, then grab the tenant (directory) id and application (client) id from there, since both appear on the same page in the App Registration and they have to go there anyway for the client id.

sadasant replied:
Thank you! I'll use that 👍

chradek commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+3. On the "Base Name", you'll see `[resourceGroup().name]`. Remove the contents of this field and type any alphanumeric string. This string will be used to generate unique names for the resources that will be created.
+4. To fill the "Tenant Id" field, copy it from the one you'll be able to see once you go to the [overview page of the Azure Active Directory section of the Azure portal](https://ms.portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/Overview).
+5. To fill the "Test Application Oid", you'll first need to create an Azure Active Directory application (Read: [How to: Use the portal to create an Azure AD application and service principal that can access resources](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal)). Once created, go to your application's overview page, and you'll see an "Ob ...
+6. Finally, accept the terms and conditions, then click on the "Purchase" button.
```
AAAHHH, purchase?! :smile: How do I delete the resources if I hit this by mistake (or just want to delete in general)?

sadasant replied:
I'll mention how to delete this!

bterlson replied:
Presumably they should READ the terms and conditions, not just accept them? 😁

bterlson commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+To fill this form:
+
+1. If you know about ARM templates, or you're interested to learn about them, we recommend you to take some time to read through the [ARM template used by the deploy button in this readme](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about ARM templates in the [Azure Resource Manager templates overview](https://do ...
+2. The subscription, resource group and location are required fields, they help specify where to create the KeyVault.
```
```suggestion
2. The subscription, resource group and location are required to specify where to create the KeyVault.
```

bterlson commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
 
+To fill this form:
+
+1. If you know about ARM templates, or you're interested to learn about them, we recommend you to take some time to read through the [ARM template used by the deploy button in this readme](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about ARM templates in the [Azure Resource Manager templates overview](https://do ...
```
I'm not sure about this bullet point - users shouldn't feel like they have to learn a super complex topic just to use the simple deployment form. Also this is the readme so we should prefer short and sweet!

sadasant replied:
Ok! What do you think if we move similar to this to the end of the bullet points? Like "If you want to learn more about.... please read...."

bterlson replied:
A link anywhere is fine by me if it's short. Like "Learn more about [arm templates]()."

sadasant replied:
I moved this to another file, just in this same branch, but this PR seems stale. Are you able to see this commit as part of this PR? https://github.com/sadasant/azure-sdk-for-js/commit/48aa4b909c27692322c1ff00a1d9eb1f2f88a436

bterlson commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+
+1. If you know about ARM templates, or you're interested to learn about them, we recommend you to take some time to read through the [ARM template used by the deploy button in this readme](https://github.com/Azure/azure-sdk-for-js/blob/keyvault/deploy-button-guide/sdk/keyvault/test-resources.json). You can read more about ARM templates in the [Azure Resource Manager templates overview](https://do ...
+2. The subscription, resource group and location are required fields, they help specify where to create the KeyVault.
+3. On the "Base Name", you'll see `[resourceGroup().name]`. Remove the contents of this field and type any alphanumeric string. This string will be used to generate unique names for the resources that will be created.
```
```suggestion
3. For "Base Name", replace `[resourceGroup().name]` with any alphanumeric string. All new resources will use this name as a prefix.
```

bterlson commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+2. The subscription, resource group and location are required fields, they help specify where to create the KeyVault.
+3. On the "Base Name", you'll see `[resourceGroup().name]`. Remove the contents of this field and type any alphanumeric string. This string will be used to generate unique names for the resources that will be created.
+4. To fill the "Tenant Id" field, copy it from the one you'll be able to see once you go to the [overview page of the Azure Active Directory section of the Azure portal](https://ms.portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/Overview).
+5. To fill the "Test Application Oid", you'll first need to create an Azure Active Directory application (Read: [How to: Use the portal to create an Azure AD application and service principal that can access resources](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal)). Once created, go to your application's overview page, and you'll see an "Ob ...
```
I can help tersify once you've made changes based on @chradek's comments 😀

bterlson commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
 
 [![](http://azuredeploy.net/deploybutton.png)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-sdk-for-js%2Fmaster%2Fsdk%2Fkeyvault%2Ftest-resources.json)
 
+To fill this form:
```
I feel like these instructions should be in a separate file? They're very long and are unlikely to be useful except maybe once, so having this be so prominent in the readme might push down other valuable content like install instructions. Really really want install instructions "above the fold".

bterlson replied:
Anyone else have feelings either way here?

chradek replied:
That seems fair, then just have a link to that separate file here.

sadasant replied:
I moved them to a separate file!

bterlson commented on sdk/keyvault/keyvault-keys/README.md:
```diff
@@ -40,6 +40,18 @@ To quickly create the needed Key Vault resources in Azure and to receive a conne
+4. To fill the "Tenant Id" field, copy it from the one you'll be able to see once you go to the [overview page of the Azure Active Directory section of the Azure portal](https://ms.portal.azure.com/#blade/Microsoft_AAD_IAM/ActiveDirectoryMenuBlade/Overview).
+5. To fill the "Test Application Oid", you'll first need to create an Azure Active Directory application (Read: [How to: Use the portal to create an Azure AD application and service principal that can access resources](https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal)). Once created, go to your application's overview page, and you'll see an "Ob ...
+6. Finally, accept the terms and conditions, then click on the "Purchase" button.
+7. A badge will appear on the top right of the portal, indicating that the resource is being created. You can go to your resource group's page, then go to the "Deployments" section to see a history of all the deployments you've made. You'll see that those deployments that have used an ARM template will say in their names "Microsoft.Template". Once you click in one of them, you'll see at the left  ...
```
```suggestion
7. A notification will appear on the top right of the portal indicating that the resource is being created. Clicking on the notification will show you deployment status. You will also find this information under your resource group's *Deployments* page.
```

* I think portal calls these things notifications not badges...
* General tutorial about managing resources seems out of scope for this document.

sadasant commented:
I can't see the latest commit I made to this branch in this PR...

chradek commented:
I also don't see the file you created in this PR. Did you do some sort of git force push, or forget to pull the changes from the branch after committing the suggestions? When I checkout your branch I notice it doesn't have the 3 recent commits that show up on GitHub (which look like committed suggestions), but I see your 'using a separate file' commit.

sadasant commented:
@chradek yeah I did force push since I rebased over master (it was a missclick) I think I manually addressed your suggestions in the other file, at least I intended to. I'm a bit concerned about the missing history here.

sadasant commented:
Closing this on favor of https://github.com/Azure/azure-sdk-for-js/pull/7432
